---
title: "5a.1 Serum Metabolites - MetaCyc - Fig 3cd"
author: "Avril Metcalfe-Roach"
date: "`r Sys.Date()`"
output: html_document
---

```{r Load data}
library(phyloseq)
library(tidyverse)
library(ggpubr)
library(microbiome)
library(ggcorrplot)
library(scales)

# Taxonomy Data - CLR transformed
ps = readRDS("../Reference Files/functional_phyloseq.rds") %>% .[["pwy"]] %>% microbiome::transform('clr')
# Significant microbially-derived metabolites from previous paper (serum metabolomics)
# p-cresol and phenylacetylglutamine
met = read.csv('../Reference Files/normalized_proteolytic_metabolites.csv')

# Load functions that allow high-throughput Spearman & GLM analysis
source('999. Cross-Sectional Correlation Functions.R')
```

```{r Prep dataset}
# Load sig pwy
temp2 = readxl::read_xlsx('../Results/Supplementary Data/Differential Abundance/MetaCyc_Summary.xlsx',sheet='multivar') %>%
  filter(sig_in_at_least_n==1, Variable=='StatusPD')

# Make a vector of the pathways to test
sig = temp2[['Species']]

# Select species and melt phyloseq object
psm = ps %>% subset_taxa(Species %in% sig) %>% psmelt() %>%
  select(Species,everything()) %>% select(-OTU) %>%
  pivot_wider(names_from = Species, values_from = Abundance)

# Combine the datasets
psm = psm %>% left_join(met)
```

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Univariate Analysis

```{r With Laxatives}
# Filter out NA values
temp = psm %>% filter(!is.na(pcresol))
spearman.lax = correlate_met(temp, 'pcresol',grp='All') %>% 
  rbind(correlate_met(temp, 'phenylacetylglutamine',grp='All')) %>% 
  rbind(correlate_met(temp %>% filter(Status=='PD'),'pcresol',grp='PD')) %>% 
  rbind(correlate_met(temp %>% filter(Status=='PD'),'phenylacetylglutamine',grp='PD')) %>% 
  rbind(correlate_met(temp %>% filter(Status=='Ctrl'),'pcresol',grp='Ctrl')) %>% 
  rbind(correlate_met(temp %>% filter(Status=='Ctrl'),'phenylacetylglutamine',grp='Ctrl'))

# Extract just the relevant rows for plotting
spearman.lax = spearman.lax %>%
  mutate(taxon = str_replace(taxon, '[&]beta;-D','B-D')) %>% 
  mutate(taxon = str_wrap(taxon,40))
```

```{r SET ORDER FOR ALL PLOTS}
# Arrange metabolites by p-cresol estimates, group 'all' ~~~~~~~~~~~~~
species_order = spearman.lax %>% filter(group=='All',variable=='pcresol') %>% 
  arrange(rho)
species_order$taxon = str_remove_all(species_order$taxon,'s__') %>% str_replace_all('_',' ')
```

```{r Fig S3 - With Laxatives}
# Edit levels for plotting
spearman.lax[spearman.lax=='pcresol'] = 'p-Cresol'
spearman.lax[spearman.lax=='phenylacetylglutamine'] = 'Phenylacetylglutamine'
spearman.lax$taxon = str_remove_all(spearman.lax$taxon,'s__') %>% str_replace_all('_',' ')

# Plotting function - labels the plot automatically
# vars = the variables you want to include in the plot
prep_plot = function(vars){
  p = spearman.lax %>% 
    mutate(Qval = ifelse(qval>0.05,'',
                        ifelse(qval>0.01,'*',
                               ifelse(qval>0.001,'**',
                                      ifelse(qval<=0.001,'***',''))))) %>% 
    mutate(Qval2 = ifelse(qval>0.1,'',
                         ifelse(qval>0.05,'+',''))) %>% 
    filter(variable %in% vars) %>% 
    mutate(variable = ifelse(variable=='Phenylacetylglutamine','Phenylacetyl-\nglutamine',variable)) %>% 
    ggplot(aes(group,taxon,col = rho)) +
    geom_point(size =14) +
    theme_classic(base_size = 16) +
    scale_color_gradient2(low="darkblue", high="darkred", guide="colorbar") +
    geom_text(aes(label=Qval),size = 10, col = 'white',nudge_y = -0.25) +
    geom_text(aes(label=Qval2),size = 10, col = 'white',nudge_y = 0) +
    ylab(NULL) + xlab(NULL) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    facet_wrap('variable')
  return(p)
}

spearman.lax$taxon = factor(spearman.lax$taxon, levels = species_order$taxon)

# Plot amino acids
p1 = prep_plot(c('Phenylacetylglutamine','p-Cresol'))

ggsave('../Results/Supplementary Figures/Fig. 3cd - MetaCyc vs Metabolites - Univar.png',height=5.5,width=8.5)
```

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Multivariate Analysis

```{r With Laxatives}
# Filter out NA values
temp = psm %>% filter(!is.na(pcresol))
glm.lax = glm_met(temp,'pcresol',grp='All') %>% 
  rbind(glm_met(temp,'phenylacetylglutamine',grp='All')) %>% 
  rbind(glm_met(temp %>% filter(Status=='PD'),'pcresol',grp='PD')) %>% 
  rbind(glm_met(temp %>% filter(Status=='PD'),'phenylacetylglutamine',grp='PD')) %>% 
  rbind(glm_met(temp %>% filter(Status=='Ctrl'),'pcresol',grp='Ctrl')) %>% 
  rbind(glm_met(temp %>% filter(Status=='Ctrl'),'phenylacetylglutamine',grp='Ctrl'))

# Extract just the relevant rows for plotting
glm_species = glm.lax %>% filter(!str_detect(Explanatory,'df[$]'), Explanatory != '(Intercept)') %>% 
  mutate(Taxon = str_replace(Taxon, '[&]beta;-D','B-D')) %>% 
  mutate(Taxon = str_wrap(Taxon,40))
```

```{r Fig S3 - With Laxatives}
# Edit levels for plotting
glm_species[glm_species=='pcresol'] = 'p-Cresol'
glm_species[glm_species=='phenylacetylglutamine'] = 'Phenylacetylglutamine'
glm_species$Explanatory = str_remove_all(glm_species$Explanatory,'s__') %>% str_replace_all('_',' ')

# Plotting function - labels the plot automatically
# vars = the variables you want to include in the plot
prep_plot = function(vars){
  p = glm_species %>% 
    mutate(Qval = ifelse(qval>0.05,'',
                        ifelse(qval>0.01,'*',
                               ifelse(qval>0.001,'**',
                                      ifelse(qval<=0.001,'***',''))))) %>% 
    mutate(Qval2 = ifelse(qval>0.1,'',
                         ifelse(qval>0.05,'+',''))) %>% 
    filter(Response %in% vars) %>% 
    mutate(Response = ifelse(Response=='Phenylacetylglutamine','Phenylacetyl-\nglutamine',Response)) %>% 
    ggplot(aes(group,Taxon,col = Estimate)) +
    geom_point(size =14) +
    theme_classic(base_size = 16) +
    scale_color_gradient2(low="darkblue", high="darkred", guide="colorbar") +
    geom_text(aes(label=Qval),size = 9, col = 'white',nudge_y = -0.15) +
    geom_text(aes(label=Qval2),size = 10, col = 'white',nudge_y = 0) +
    ylab(NULL) + xlab(NULL) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    facet_wrap('Response')
  return(p)
}

glm_species$Taxon = factor(glm_species$Taxon, levels = species_order$taxon)

# Plot amino acids
p1 = prep_plot(c('Phenylacetylglutamine','p-Cresol'))
ggsave('../Results/3. Metabolite Associations/Fig. 3cd - MetaCyc vs Metabolites - Multivar.png',height=5.5,width=8.5)
```

# Plot some key findings

```{r}
p1 = psm %>% ggplot(aes(`PWY-7383: anaerobic energy metabolism (invertebrates, cytosol)`,pcresol,col=Status)) +
  geom_point() + geom_smooth(method='lm') +
  theme_classic(base_size=18) +
  ggpubr::stat_cor(method='spearman',cor.coef.name = 'rho',size=6) +
  ylab('p-Cresol') + xlab('Anaerobic Energy Metabolism')

p2 = psm %>% ggplot(aes(`PWY-7383: anaerobic energy metabolism (invertebrates, cytosol)`,phenylacetylglutamine,col=Status)) +
  geom_point() + geom_smooth(method='lm') +
  theme_classic(base_size=18) +
  ggpubr::stat_cor(method='spearman',cor.coef.name = 'rho',size=5.5) +
  ylab('Phenylacetylglutamine') + xlab('Anaerobic Energy Metabolism')

ggarrange(plotlist = list(p1,p2), common.legend = T, legend = 'right')
ggsave('../Results/3. Metabolite Associations/Fig. 3d - MetaCyc Scatterplots.jpeg',height=4,width=9)
```

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```{r Save Stats}
writexl::write_xlsx(list('Univar' = spearman.lax,
                         'Multivar' = glm.lax,
                         'Multivar Species' = glm.lax %>% filter(str_detect(Explanatory,'Microbe'))), '../Results/Supplementary Data/Metabolite Associations/MetaCyc vs Metabolites.xlsx')
```