---
title: "5a.2 Serum Metabolites - COG"
author: "Avril Metcalfe-Roach"
date: "`r Sys.Date()`"
output: html_document
---

```{r Load data}
library(phyloseq)
library(tidyverse)
library(ggpubr)
library(microbiome)
library(ggcorrplot)
library(scales)

# Taxonomy Data - CLR transformed
ps = readRDS("../Reference Files/functional_phyloseq.rds") %>% .[["eggnog"]] %>% microbiome::transform('clr')
# Significant microbially-derived metabolites from previous paper (serum metabolomics)
# p-cresol and phenylacetylglutamine
met = read.csv('../Reference Files/normalized_proteolytic_metabolites.csv')

# Load functions that allow high-throughput Spearman & GLM analysis
source('999. Cross-Sectional Correlation Functions.R')
```

```{r Prep dataset}
# Load sig pwy
temp = readxl::read_xlsx('../Results/Supplementary Data/Differential Abundance/COG Enzymes_Stats2.xlsx',sheet='univar') %>% 
  filter(q_val<0.05) %>% group_by(Species) %>% add_count() %>% filter(n>1) %>% 
  select(Species,association_dir) %>% unique
temp2 = readxl::read_xlsx('../Results/Supplementary Data/Differential Abundance/COG Enzymes_Stats2.xlsx',sheet='multivar') %>%
  filter(Variable=='StatusPD', Species %in% temp$Species) %>% 
  group_by(DA) %>% mutate(q_val = p.adjust(p_val,method='BH')) %>% ungroup %>% 
  filter(q_val<0.05) %>% group_by(Species) %>% add_count() %>% filter(n>1) %>% 
  select(Species,association_dir) %>% unique

# Make a vector of the pathways to test
sig = temp2[['Species']]

# Select species and melt phyloseq object
psm = ps %>% subset_taxa(Species %in% sig) %>% psmelt() %>%
  select(Species,everything()) %>% select(-OTU) %>%
  pivot_wider(names_from = Species, values_from = Abundance)

# Combine the datasets
psm = psm %>% left_join(met)
```

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Univariate Analysis

```{r}
# Filter out NA values
temp = psm %>% filter(!is.na(pcresol))
spearman.lax = correlate_met(temp, 'pcresol',grp='All') %>% 
  rbind(correlate_met(temp, 'phenylacetylglutamine',grp='All')) %>% 
  rbind(correlate_met(temp %>% filter(Status=='PD'),'pcresol',grp='PD')) %>% 
  rbind(correlate_met(temp %>% filter(Status=='PD'),'phenylacetylglutamine',grp='PD')) %>% 
  rbind(correlate_met(temp %>% filter(Status=='Ctrl'),'pcresol',grp='Ctrl')) %>% 
  rbind(correlate_met(temp %>% filter(Status=='Ctrl'),'phenylacetylglutamine',grp='Ctrl'))

# Extract just the relevant rows for plotting
spearman.lax = spearman.lax %>%
  mutate(taxon = str_wrap(taxon,40))
```

```{r SET ORDER FOR ALL PLOTS}
# Arrange metabolites by p-cresol estimates, group 'all' ~~~~~~~~~~~~~
species_order = spearman.lax %>% filter(group=='All',variable=='pcresol') %>% 
  arrange(rho)
```

```{r}
# Edit levels for plotting
spearman.lax[spearman.lax=='pcresol'] = 'p-Cresol'
spearman.lax[spearman.lax=='phenylacetylglutamine'] = 'Phenylacetylglutamine'

# Plotting function - labels the plot automatically
# vars = the variables you want to include in the plot
prep_plot = function(vars){
  p = spearman.lax %>% 
    mutate(Qval = ifelse(qval>0.05,'',
                        ifelse(qval>0.01,'*',
                               ifelse(qval>0.001,'**',
                                      ifelse(qval<=0.001,'***',''))))) %>% 
    mutate(Qval2 = ifelse(qval>0.1,'',
                         ifelse(qval>0.05,'+',''))) %>% 
    filter(variable %in% vars) %>% 
    mutate(variable = ifelse(variable=='Phenylacetylglutamine','Phenylacetyl-\nglutamine',variable)) %>% 
    ggplot(aes(group,taxon,col = rho)) +
    geom_point(size =14) +
    theme_classic(base_size = 16) +
    scale_color_gradient2(low="darkblue", high="darkred", guide="colorbar") +
    geom_text(aes(label=Qval),size = 10, col = 'white',nudge_y = -0.25) +
    geom_text(aes(label=Qval2),size = 10, col = 'white',nudge_y = 0) +
    ylab(NULL) + xlab(NULL) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    facet_wrap('variable')
  return(p)
}

spearman.lax$taxon = factor(spearman.lax$taxon, levels = species_order$taxon)

# Plot amino acids
p1 = prep_plot(c('Phenylacetylglutamine','p-Cresol'))
ggsave('../Results/Supplementary Figures/Fig. 3cd - COG vs Metabolites - Univar.png',height=20,width=8)
```

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Multivariate Analysis

```{r}
# Filter out NA values
temp = psm %>% filter(!is.na(pcresol))
glm.lax = glm_met(temp,'pcresol',grp='All') %>% 
  rbind(glm_met(temp,'phenylacetylglutamine',grp='All')) %>% 
  rbind(glm_met(temp %>% filter(Status=='PD'),'pcresol',grp='PD')) %>% 
  rbind(glm_met(temp %>% filter(Status=='PD'),'phenylacetylglutamine',grp='PD')) %>% 
  rbind(glm_met(temp %>% filter(Status=='Ctrl'),'pcresol',grp='Ctrl')) %>% 
  rbind(glm_met(temp %>% filter(Status=='Ctrl'),'phenylacetylglutamine',grp='Ctrl'))

# Extract just the relevant rows for plotting
glm_species = glm.lax %>% filter(!str_detect(Explanatory,'df[$]'), Explanatory != '(Intercept)') %>% 
  mutate(Taxon = str_wrap(Taxon,40))
```

```{r}
# Edit levels for plotting
glm_species[glm_species=='pcresol'] = 'p-Cresol'
glm_species[glm_species=='phenylacetylglutamine'] = 'Phenylacetylglutamine'

# Plotting function - labels the plot automatically
# vars = the variables you want to include in the plot
prep_plot = function(vars){
  p = glm_species %>% 
    mutate(Qval = ifelse(qval>0.05,'',
                        ifelse(qval>0.01,'*',
                               ifelse(qval>0.001,'**',
                                      ifelse(qval<=0.001,'***',''))))) %>% 
    mutate(Qval2 = ifelse(qval>0.1,'',
                         ifelse(qval>0.05,'+',''))) %>% 
    filter(Response %in% vars) %>% 
    mutate(Response = ifelse(Response=='Phenylacetylglutamine','Phenylacetyl-\nglutamine',Response)) %>% 
    ggplot(aes(group,Taxon,col = Estimate)) +
    geom_point(size =14) +
    theme_classic(base_size = 14) +
    scale_color_gradient2(low="darkblue", high="darkred", guide="colorbar") +
    geom_text(aes(label=Qval),size = 9, col = 'white',nudge_y = -0.15) +
    geom_text(aes(label=Qval2),size = 10, col = 'white',nudge_y = 0) +
    ylab(NULL) + xlab(NULL) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    facet_wrap('Response')
  return(p)
}

glm_species$Taxon = factor(glm_species$Taxon, levels = species_order$taxon)

# Plot amino acids
p1 = prep_plot(c('Phenylacetylglutamine','p-Cresol'))

ggsave('../Results/Supplementary Figures/Fig. 3cd - COG vs Metabolites - Multivar.png',height=20,width=8)
```

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```{r Save Stats}
writexl::write_xlsx(list('Univar' = spearman.lax,
                         'Multivar' = glm.lax,
                         'Multivar Species' = glm.lax %>% filter(str_detect(Explanatory,'Microbe'))),                    
                    '../Results/Supplementary Data/Metabolite Associations/COG vs Metabolites.xlsx')
```